### PR TITLE
Settings: Refactor Post Formats to use `react-query`

### PR DESCRIPTION
**Pull request by tyxla**
#### Changes proposed in this Pull Request

Considering that post formats a very straightforward example of server state on the client side in Calypso, and the simplicity of the use case (read-only), this makes it a great example of how much boilerplate we can remove by using `react-query` instead of redux.

This PR refactors the post formats Redux state to use `react-query` and ports the one single usage of this state tree.

#### Testing instructions

* Make sure your WP.com site is using a theme that supports one or more custom post formats. You can find such a theme here: https://wordpress.com/themes/filter/post-formats/:site where `:site` is your WP.com site.
* Go to `/settings/writing/:site`
* Verify the post formats in the "Default post format" are loading correctly.
* Pick a site with a theme without any custom post formats. 
* Verify the field correctly displays just the "Standard" option and there are no errors in the console.

### CODE REVIEW

File: client/data/post-formats/use-post-formats-query.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import { useQuery } from 'react-query';
+
+/**
+ * Internal dependencies
+ */
+import wpcom from 'calypso/lib/wp';
+
+const usePostFormatsQuery = ( siteId ) =>
+	useQuery( [ 'postFormats', siteId ], () =>
+		wpcom.undocumented().site( siteId ).postFormatsList()
```
Comment (flootr) [inline client/data/post-formats/use-post-formats-query.js]: I think we started using `wpcom.req.get` directly so we can eventually move away from `wpcom.undocumented`.

File: client/my-sites/site-settings/composing/default-post-format.jsx [hunk 2]
```diff
@@ -42,14 +39,12 @@ const DefaultPostFormat = ( {
...
-				{ postFormats &&
-					map( postFormats, ( label, slug ) => {
-						return (
-							<option key={ slug } value={ slug }>
-								{ label }
-							</option>
-						);
-					} ) }
+				{ data &&
+					map( data.formats, ( label, slug ) => (
```
Comment (flootr) [inline client/my-sites/site-settings/composing/default-post-format.jsx]: Can we avoid using `_.map` in this case? I know it's preexisting to this PR but may be a good side-refactor.

File: client/my-sites/site-settings/composing/default-post-format.jsx [hunk 3]
```diff
@@ -69,11 +64,4 @@ DefaultPostFormat.propTypes = {
...
 
-export default connect( ( state ) => {
-	const siteId = getSelectedSiteId( state );
-
-	return {
-		siteId,
-		postFormats: getPostFormats( state, siteId ),
-	};
-} )( localize( DefaultPostFormat ) );
+export default localize( DefaultPostFormat );
```
Comment (flootr) [inline client/my-sites/site-settings/composing/default-post-format.jsx]: As this is now a functional component what about using `useTranslate` instead?

File: client/my-sites/site-settings/composing/default-post-format.jsx [hunk 4]
```diff
@@ -1,37 +1,34 @@
...
 	fields,
 	onChangeField,
 	eventTracker,
 	isRequestingSettings,
 	isSavingSettings,
-	postFormats,
-	siteId,
 	translate,
 } ) => {
+	const siteId = useSelector( ( state ) => getSelectedSiteId( state ) );
```
Comment (flootr) [inline client/my-sites/site-settings/composing/default-post-format.jsx]: nit: could this be point-free `useSelector( getSelectedSiteId )` or would it have unintentional side effects?

**Review by flootr — COMMENTED** [review]
I ❤️ this. Just a few comments. Let me know what you think.

File: client/data/post-formats/use-post-formats-query.js (hunk 1 above)
Comment (tyxla) [reply client/data/post-formats/use-post-formats-query.js]: Sure thing! Done in fa174d5bc515b11810f2d9d1a09f91731dfb8ddf. I've also removed the unused method in a0aad533a2820183e474c9feb75bdcf60b4e0c67.

File: client/my-sites/site-settings/composing/default-post-format.jsx (hunk 4 above)
Comment (tyxla) [reply client/my-sites/site-settings/composing/default-post-format.jsx]: Sure 👍  Done in fa174d5bc515b11810f2d9d1a09f91731dfb8ddf

File: client/my-sites/site-settings/composing/default-post-format.jsx (hunk 2 above)
Comment (tyxla) [reply client/my-sites/site-settings/composing/default-post-format.jsx]: I don't mind refactoring that as it's the only Lodash usage here. Removed in 36d2ed21465999c528175be8d95fa819d06f0d63.

File: client/my-sites/site-settings/composing/default-post-format.jsx (hunk 3 above)
Comment (tyxla) [reply client/my-sites/site-settings/composing/default-post-format.jsx]: 👍  Done in 976158ac1266a37c058c203e938cde637c3a3e19.

**Comment by tyxla** [comment]
> I ❤️ this. Just a few comments. Let me know what you think.

Thanks for the great review, @flootr! I've addressed all the feedback, could you take another look?

**Review by flootr — APPROVED** [review]
This works great 👍🏻 awesome refactor!

There's two things I want to mention though (I don't consider any of them as blocking):

- It seems like post formats used to be persisted (with schema validation), that's not the case with ReactQuery anymore (we haven't implemented this yet). However, I'm not sure if that's really needed for this kind of data. Looping in @jsnajdr as well as we should discuss on how to move on there.
- I'm not sure about the level of "freshness" we need for post formats. Right now data is refetched on focus change and its stale time is very low (that's the default). Post formats only change when the active theme is switched, correct? We could increase `staleTime` and/or disable `refetchOnWindowFocus` but I'm not sure about what works best and I guess this is something we should to iterate on in the future.

**Comment by tyxla** [comment]
> It seems like post formats used to be persisted (with schema validation), that's not the case with ReactQuery anymore (we haven't implemented this yet). However, I'm not sure if that's really needed for this kind of data. Looping in @jsnajdr as well as we should discuss on how to move on there.

I don't think it's needed, TBH. I don't see much value in it, given how simple the use case is.

> I'm not sure about the level of "freshness" we need for post formats. Right now data is refetched on focus change and its stale time is very low (that's the default). Post formats only change when the active theme is switched, correct? We could increase staleTime and/or disable refetchOnWindowFocus but I'm not sure about what works best and I guess this is something we should to iterate on in the future.

I've been thinking about that and decided to keep the defaults, because that way if the user changes the theme and goes back to that window, they will be refetched. However, I'm happy to disable those if you think that default freshness is unnecessary and only adds extra request overhead.

That being said, I'll refrain from shipping this for now until we have a consensus here - let me know what you think @flootr @jsnajdr.

**Comment by flootr** [comment]
@tyxla from my side this is ready to go, my comments were meant as general notes.

**Comment by tyxla** [comment]
I'll go ahead and ship this then. Happy to follow up if you folks have suggestions.

Thanks again, Markus 🙌

**Pull request merged**